File: src/logic/navmesh/NavUtils.ts
import type { Navmesh } from "./Navmesh";
import { Point2 } from "../core/math";
import { seededRandom } from "../core/mathUtils";


export function findTriangle(point: Point2, navmesh: Navmesh, lastTriangle: number): number {
    const checkTriangle = (triIdx: number) => {
        if (triIdx < 0 || triIdx * 3 >= navmesh.triangles.length) {
            return false;
        }
        
        return testPointInsideTriangle(navmesh, point.x, point.y, triIdx);
    };
    
    // Try the last triangle first if provided
    if (lastTriangle >= 0) {
        if (checkTriangle(lastTriangle)) {
            return lastTriangle;
        }
    }
    const candidateTriangles = navmesh.triangleIndex.query(point.x, point.y);
    
    for (const triIdx of candidateTriangles) {
        if (checkTriangle(triIdx)) {
            return triIdx;
        }
    }
    return -1;
}

export function isPointInNavmesh(point: Point2, navmesh: Navmesh, lastTriangle: number): number {
    const triIdx = findTriangle(point, navmesh, lastTriangle);
    if (triIdx === -1 || triIdx >= navmesh.walkable_triangle_count) {
        return -1;
    }
    return triIdx;
}

export function getRandomTriangle(navmesh: Navmesh, seed?: number): number {
    let currentSeed = seed ?? 12345;
    const maxAttempts = 10;
    
    for (let i = 0; i < maxAttempts; i++) {
        const { value: randomX, newSeed: newSeedX } = seededRandom(currentSeed);
        currentSeed = newSeedX;
        const { value: randomY, newSeed: newSeedY } = seededRandom(currentSeed);
        currentSeed = newSeedY;
        
        const x = navmesh.triangleIndex.minX + randomX * (navmesh.triangleIndex.maxX - navmesh.triangleIndex.minX);
        const y = navmesh.triangleIndex.minY + randomY * (navmesh.triangleIndex.maxY - navmesh.triangleIndex.minY);
        const point: Point2 = { x, y };
        const triIndex = isPointInNavmesh(point, navmesh, -1);
        if (triIndex !== -1) {
            return triIndex;
        }
    }

    const numTriangles = navmesh.triangles.length / 3;
    const { value } = seededRandom(currentSeed);
    return Math.floor(value * numTriangles);
}

export function getRandomTriangleInArea(navmesh: Navmesh, centerX: number, centerY: number, numCellExtents: number, seed?: number): number {
    let currentSeed = seed ?? 12345;
    const maxAttempts = 20;
    
    const halfExtent = numCellExtents * navmesh.triangleIndex.cellSize;
    const minX = centerX - halfExtent;
    const maxX = centerX + halfExtent;
    const minY = centerY - halfExtent;
    const maxY = centerY + halfExtent;
    
    const clampedMinX = Math.max(minX, navmesh.triangleIndex.minX);
    const clampedMaxX = Math.min(maxX, navmesh.triangleIndex.maxX);
    const clampedMinY = Math.max(minY, navmesh.triangleIndex.minY);
    const clampedMaxY = Math.min(maxY, navmesh.triangleIndex.maxY);
    
    for (let i = 0; i < maxAttempts; i++) {
        const { value: randomX, newSeed: newSeedX } = seededRandom(currentSeed);
        currentSeed = newSeedX;
        const { value: randomY, newSeed: newSeedY } = seededRandom(currentSeed);
        currentSeed = newSeedY;
        
        const x = clampedMinX + randomX * (clampedMaxX - clampedMinX);
        const y = clampedMinY + randomY * (clampedMaxY - clampedMinY);
        const point: Point2 = { x, y };
        const triIndex = isPointInNavmesh(point, navmesh, -1);
        if (triIndex !== -1) {
            return triIndex;
        }
    }

    const startCellX = Math.max(0, Math.floor((clampedMinX - navmesh.triangleIndex.minX) / navmesh.triangleIndex.cellSize));
    const endCellX = Math.min(navmesh.triangleIndex.gridWidth - 1, Math.floor((clampedMaxX - navmesh.triangleIndex.minX) / navmesh.triangleIndex.cellSize));
    const startCellY = Math.max(0, Math.floor((clampedMinY - navmesh.triangleIndex.minY) / navmesh.triangleIndex.cellSize));
    const endCellY = Math.min(navmesh.triangleIndex.gridHeight - 1, Math.floor((clampedMaxY - navmesh.triangleIndex.minY) / navmesh.triangleIndex.cellSize));
    
    const candidateTriangles: number[] = [];
    for (let cx = startCellX; cx <= endCellX; cx++) {
        for (let cy = startCellY; cy <= endCellY; cy++) {
            const cellTriangles = getTrianglesInCell(navmesh, cx, cy);
            for (const triIdx of cellTriangles) {
                if (!candidateTriangles.includes(triIdx)) {
                    candidateTriangles.push(triIdx);
                }
            }
        }
    }
    
    if (candidateTriangles.length > 0) {
        const { value } = seededRandom(currentSeed);
        const randomIndex = Math.floor(value * candidateTriangles.length);
        return candidateTriangles[randomIndex];
    }
    
    return getRandomTriangle(navmesh, currentSeed);
}


export function getTrianglesInCell(navmesh: Navmesh, cx: number, cy: number): Int32Array {
    const cellIndex = cx + cy * navmesh.triangleIndex.gridWidth;
    if (cellIndex < 0 || cellIndex >= navmesh.triangleIndex.cellOffsets.length - 1) {
        return new Int32Array(0);
    }
    const start = navmesh.triangleIndex.cellOffsets[cellIndex];
    const end = navmesh.triangleIndex.cellOffsets[cellIndex + 1];
    return navmesh.triangleIndex.cellItems.slice(start, end);
}

export function testPointInsideTriangle(navmesh: Navmesh, x: number, y: number, tri_idx: number): boolean {
    // Get the three vertex indices for this triangle
    const v1_idx = navmesh.triangles[tri_idx * 3];
    const v2_idx = navmesh.triangles[tri_idx * 3 + 1]; 
    const v3_idx = navmesh.triangles[tri_idx * 3 + 2];
    
    // Get the actual coordinates of the vertices
    const ax = navmesh.vertices[v1_idx * 2];
    const ay = navmesh.vertices[v1_idx * 2 + 1];
    const bx = navmesh.vertices[v2_idx * 2];
    const by = navmesh.vertices[v2_idx * 2 + 1];
    const cx = navmesh.vertices[v3_idx * 2];
    const cy = navmesh.vertices[v3_idx * 2 + 1];
    
    // Edge v1-v2: ensure canonical order (smaller index first)
    const flip12 = v1_idx > v2_idx;
    const o12 = flip12 
        ? (ax - bx) * (y - by) - (ay - by) * (x - bx)  // b->a
        : (bx - ax) * (y - ay) - (by - ay) * (x - ax); // a->b
    if (!(flip12 ? (o12 <= 0) : (o12 >= 0))) return false;
    
    // Edge v2-v3: ensure canonical order (smaller index first)  
    const flip23 = v2_idx > v3_idx;
    const o23 = flip23
        ? (bx - cx) * (y - cy) - (by - cy) * (x - cx)  // c->b
        : (cx - bx) * (y - by) - (cy - by) * (x - bx); // b->c
    if (!(flip23 ? (o23 <= 0) : (o23 >= 0))) return false;
    
    // Edge v3-v1: ensure canonical order (smaller index first)
    const flip31 = v3_idx > v1_idx;
    const o31 = flip31
        ? (cx - ax) * (y - ay) - (cy - ay) * (x - ax)  // a->c
        : (ax - cx) * (y - cy) - (ay - cy) * (x - cx); // c->a
    
    return flip31 ? (o31 <= 0) : (o31 >= 0);
}

export function testPointInsidePolygon(navmesh: Navmesh, x: number, y: number, poly_idx: number): boolean {
    const poly_start = navmesh.polygons[poly_idx];
    const poly_end = navmesh.polygons[poly_idx + 1];
    const poly_vert_count = poly_end - poly_start;

    for (let i = 0; i < poly_vert_count - 1; i++) {
        const v1_idx = navmesh.poly_verts[poly_start + i];
        const v2_idx = navmesh.poly_verts[poly_start + i + 1];

        const v1x = navmesh.vertices[v1_idx * 2];
        const v1y = navmesh.vertices[v1_idx * 2 + 1];
        const v2x = navmesh.vertices[v2_idx * 2];
        const v2y = navmesh.vertices[v2_idx * 2 + 1];

        const flip = v1_idx > v2_idx;
        const orientation = flip
            ? -((v1x - v2x) * (y - v2y) - (v1y - v2y) * (x - v2x))
            : (v2x - v1x) * (y - v1y) - (v2y - v1y) * (x - v1x);
        
        if (orientation < 0) {
            return false;
        }
    }

    const v1_idx = navmesh.poly_verts[poly_start + poly_vert_count - 1];
    const v2_idx = navmesh.poly_verts[poly_start + 0];
    
    const v1x = navmesh.vertices[v1_idx * 2];
    const v1y = navmesh.vertices[v1_idx * 2 + 1];
    const v2x = navmesh.vertices[v2_idx * 2];
    const v2y = navmesh.vertices[v2_idx * 2 + 1];
    
    const flip = v1_idx > v2_idx;
    const orientation = flip
        ? -((v1x - v2x) * (y - v2y) - (v1y - v2y) * (x - v2x))
        : (v2x - v1x) * (y - v1y) - (v2y - v1y) * (x - v1x);
    
    return orientation >= 0;
}

File: src/wasm/nav_utils.cpp
#include "nav_utils.h"
#include "math_utils.h"
#include "navmesh.h"
#include <algorithm>
#include <cmath>
#include <vector>
#include <iostream> // Added for logging

// External reference to global navmesh
extern Navmesh g_navmesh;

static bool check_triangle(int32_t triIdx, const Point2& point) {
    if (triIdx < 0 || triIdx >= g_navmesh.walkable_triangle_count) {
        return false;
    }
    
    return test_point_inside_triangle(point, triIdx);
}

int32_t is_point_in_navmesh(Point2 p, int32_t lastTriangle) {
    // Try the last triangle first if provided
    if (lastTriangle >= 0 && lastTriangle < g_navmesh.walkable_triangle_count) {
        if (check_triangle(lastTriangle, p)) {
            return lastTriangle;
        }
    }

    std::vector<int> candidateTriangles = g_navmesh.triangle_index.query(p);
    for (int32_t triIdx : candidateTriangles) {
        if (check_triangle(triIdx, p)) {
            return triIdx;
        }
    }
    return -1;
}

int32_t get_random_triangle(uint64_t* seed) {
    uint64_t local_seed = *seed;
    const int32_t maxAttempts = 10;
    
    for (int i = 0; i < maxAttempts; i++) {
        float randomX = math::seed_to_random_no_advance(&local_seed);
        float randomY = math::seed_to_random_no_advance(&local_seed);
        
        float x = g_navmesh.triangle_index.minX + randomX * (g_navmesh.triangle_index.maxX - g_navmesh.triangle_index.minX);
        float y = g_navmesh.triangle_index.minY + randomY * (g_navmesh.triangle_index.maxY - g_navmesh.triangle_index.minY);
        
        int32_t triIndex = is_point_in_navmesh({x, y}, -1);
        if (triIndex != -1) {
            return triIndex;
        }
    }
    
    // Fallback: return a random triangle index
    const int32_t numTriangles = g_navmesh.walkable_triangle_count;
    if (numTriangles > 0) {
        float randomValue = math::seed_to_random_no_advance(&local_seed);
        return static_cast<int32_t>(randomValue * numTriangles);
    }
    
    return -1;
}

int32_t get_random_triangle_in_area(Point2 center, int32_t numCellExtents, uint64_t* seed) {
    uint64_t local_seed = *seed;
    const int32_t maxAttempts = 20;
    
    const float halfExtent = numCellExtents * g_navmesh.triangle_index.cellSize;
    const float minX = center.x - halfExtent;
    const float maxX = center.x + halfExtent;
    const float minY = center.y - halfExtent;
    const float maxY = center.y + halfExtent;
    
    const float clampedMinX = std::max(minX, g_navmesh.triangle_index.minX);
    const float clampedMaxX = std::min(maxX, g_navmesh.triangle_index.maxX);
    const float clampedMinY = std::max(minY, g_navmesh.triangle_index.minY);
    const float clampedMaxY = std::min(maxY, g_navmesh.triangle_index.maxY);
    
    // Try random points in the area
    for (int i = 0; i < maxAttempts; i++) {
        float randomX = math::seed_to_random_no_advance(&local_seed);
        float randomY = math::seed_to_random_no_advance(&local_seed);
        
        float x = clampedMinX + randomX * (clampedMaxX - clampedMinX);
        float y = clampedMinY + randomY * (clampedMaxY - clampedMinY);
        
        Point2 point = {x, y};
        int32_t triIndex = is_point_in_navmesh(point, -1);
        if (triIndex != -1) {
            return triIndex;
        }
    }
    
    // Fallback: collect triangles from cells in the area
    std::vector<int32_t> candidateTriangles = g_navmesh.triangle_index.queryArea(clampedMinX, clampedMinY, clampedMaxX, clampedMaxY);
    
    if (!candidateTriangles.empty()) {
        float randomValue = math::seed_to_random_no_advance(&local_seed);
        int32_t randomIndex = static_cast<int32_t>(randomValue * candidateTriangles.size());
        return candidateTriangles[randomIndex];
    }
    
    // Final fallback
    return get_random_triangle(&local_seed);
}

int32_t get_triangles_in_cell(int32_t cellX, int32_t cellY, int32_t* triangleIds, int32_t maxTriangles) {
    const int32_t cellIndex = cellX + cellY * g_navmesh.triangle_index.gridWidth;
    
    if (cellIndex < 0 || cellIndex >= static_cast<int32_t>(g_navmesh.triangle_index.cellOffsetsCount) - 1) {
        return 0;
    }
    
    const uint32_t start = g_navmesh.triangle_index.cellOffsets[cellIndex];
    const uint32_t end = g_navmesh.triangle_index.cellOffsets[cellIndex + 1];
    
    const int32_t count = std::min(maxTriangles, static_cast<int32_t>(end - start));
    
    for (int32_t i = 0; i < count; i++) {
        triangleIds[i] = g_navmesh.triangle_index.cellItems[start + i];
    }
    
    return count;
}

int getTriangleFromPoint(const Point2& point) {
    std::vector<int> possibleTris = g_navmesh.triangle_index.query(point);

    for (int triIdx : possibleTris) {
        if (test_point_inside_triangle(point, triIdx)) {
            return triIdx;
        }
    }

    return -1;
}

int getPolygonFromPoint(const Point2& point) {
    std::vector<int> possiblePolys = g_navmesh.polygon_index.query(point);
    for (int polyIdx : possiblePolys) {
        if (test_point_inside_poly_t(point, polyIdx)) {
            return polyIdx;
        }
    }
    return -1;
}

int getBlobFromPoint(const Point2& point) {
    std::vector<int> possibleBlobs = g_navmesh.blob_index.query(point);
    for (int blobIdx : possibleBlobs) {
        if (testPointInsideBlob(point, blobIdx)) {
            return blobIdx;
        }
    }
    return -1;
}

int getTriangleFromPolyPoint(const Point2& point, int poly_idx) {
    const int32_t poly_start = g_navmesh.poly_tris[poly_idx];
    const int32_t poly_end = g_navmesh.poly_tris[poly_idx + 1];

    for (int i = poly_start; i < poly_end; i++) {
        if (test_point_inside_triangle(point, i)) {
            return i;
        }
    }

    return -1;
}
